### PR TITLE
feat(Systest): enable configuration override within .test file

### LINKED ISSUE #835: [TASK] Systest: enable configuration override within .test file
### Description

Some tests require running them with multiple configurations, e.g., we want to run with specific edge cases of query engine settings, or run queries with and without query optimizer enabled and ensure that they have the same result.

Thus, we want to enable the override of single-node-worker configurations within the systest file.
I.e. the systest fiel author can specify the new configuration via:

``` SQL
# name: map/Map_Div.test
# description: Simple map tests with a div function
# groups: [Map]

Source stream INT64 id INT64 value INT64 timestamp
5,1,1000
10,1,1001
15,1,1002
20,2,2000
25,19,19000
30,20,20000
35,21,21000

SINK streamSink INT64 stream$id INT64 stream$value INT64 stream$timestamp
SINK streamSinkWithNew INT64 stream$id INT64 stream$value INT64 stream$timestamp INT64 stream$new

# --> Runs the following query once with page size 8 and once with page size 1024
Configuration worker.queryOptimizer.pageSize: [8, 1024]

SELECT id / 5 AS id, value, timestamp FROM stream INTO streamSink
----
1,1,1000
2,1,1001
3,1,1002
4,2,2000
5,19,19000
6,20,20000
7,21,21000
```

### Acceptance Criteria

- [ ] test example above works

### PR DESCRIPTION
**Pull request by Philipp-Timm**
## Purpose of the Change and Brief Change Log
This pull request makes it possible to override the configuration of the worker for specific queries within the .test file itself.

- added Configuration & GlobalConfiguration token to SystestParser
- SystestParser reads and creates copies of queries, per configOverride value.
- The syntax is: 
`Configuration <option>: [<value>, <value>, <value>]`
`Configuration <option>: <value>` (for next query)
`GlobalConfiguration worker.queryOptimizer.pageSize: [8, 1024]` ('global' keyword for all queries in file)
- changed total finished query counting behavior to work with multiple runQueries() calls
- displays the configuration overrides in the result output (i.e. whether a query failed or passed) like this:
2/8 Map_Div:05 [Override: worker.queryOptimizer.pageSize=1024].....................................PASSED

## Issue Closed by this pull request: 

This PR closes #835

### CODE REVIEW

**Review by ls-1801 — CHANGES_REQUESTED** [review]
1. Please add a few examples, if you don't use this feature it will be broken the very next commit.

The `BasicTypeSupport.test` is a good candidate, as it requires a specific buffer size.

2. I think there is a flaw in the logic. When combining global and local
```
# name: formatter/BasicTypeSupport.test
# description: formatter tests
# groups: [Formatter]

# NOTE: assumes that raw tuple buffers are of page size (4096 bytes)
GlobalConfiguration worker.bufferSizeInBytes: [4096, 8192, 16384]
GlobalConfiguration worker.defaultQueryExecution.executionMode: [INTERPRETER, COMPILER]

Source multiBuffer BOOLEAN field_1 CHAR field_2 INT8 field_3 INT16 field_4 INT32 field_5 INT64 field_6 UINT8 field_7 UINT16 field_8 UINT32 field_9 UINT64 field_10 FLOAT32 field_11 FLOAT64 field_12 INLINE
true,A,-128,-32768,-2147483648,-9223372036854775808,255,65535,4294967295,18446744073709551615,-340282346638528859811704183484516925440.0,-179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.0
false,a,127,32767,2147483647,9223372036854775807,0,0,0,0,340282346638528859811704183484516925440.0,179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.0

SINK sinkMultiBuffer BOOLEAN multiBuffer$field_1 CHAR multiBuffer$field_2 INT8 multiBuffer$field_3 INT16 multiBuffer$field_4 INT32 multiBuffer$field_5 INT64 multiBuffer$field_6 UINT8 multiBuffer$field_7 UINT16 multiBuffer$field_8 UINT32 multiBuffer$field_9 UINT64 multiBuffer$field_10 FLOAT32 multiBuffer$field_11 FLOAT64 multiBuffer$field_12

Configuration worker.bufferSizeInBytes: [32768]
# Description: Check that the formatter supports all basic types.
# Sets: true/false for bool, lower/uppercase for char and min/max values for integers/floats
SELECT * FROM multiBuffer INTO sinkMultiBuffer
----
1,A,-128,-32768,-2147483648,-9223372036854775808,255,65535,4294967295,18446744073709551615,-340282346638528859811704183484516925440.0,-179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.0
0,a,127,32767,2147483647,9223372036854775807,0,0,0,0,340282346638528859811704183484516925440.0,179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368.0
```

will execute 6 tests:
```
  1/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=COMPILER, worker.bufferSizeInBytes=32768]................PASSED 
  2/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=COMPILER, worker.bufferSizeInBytes=32768]................PASSED 
  3/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=COMPILER, worker.bufferSizeInBytes=32768]................PASSED 
  4/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=INTERPRETER, worker.bufferSizeInBytes=32768].............PASSED 
  5/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=INTERPRETER, worker.bufferSizeInBytes=32768].............PASSED 
  6/6 BasicTypeSupport:01 [worker.defaultQueryExecution.executionMode=INTERPRETER, worker.bufferSizeInBytes=32768].............PASSED 
```

3. The interaction between the old global-global configuration (i.e., when calling systest -- --worker.numberOfBuffersInGlobalBufferManager=1000) is I bit unclear to me. And I am not sure what the best solution would be. 
I think it is problematic, that running `systest -- --worker.defaultQueryExecution.executionMode=INTERPERTER` with a systest that intern overwrites worker.defaultQueryExecution.executionMode: [INTERPERTER, COMPILER], will run COMPILER tests.

File: nes-systests/systest/src/SystestRunner.cpp [hunk 1]
```diff
@@ -320,22 +319,35 @@ std::vector<RunningQuery> runQueriesAndBenchmark(
...
-    const auto queryCounterAsString = std::to_string(queryCounter + 1);
-
-    /// spd logger cannot handle multiline prints with proper color and pattern.
-    /// And as this is only for test runs we use stdout here.
-    std::cout << std::string(padSizeQueryCounter - queryCounterAsString.size(), ' ');
-    std::cout << queryCounterAsString << "/" << totalQueries << " ";
-    std::cout << runningQuery.systestQuery.testName << ":" << std::string(padSizeQueryNumber - queryNumberLength, '0')
-              << queryNumberAsString;
-    std::cout << std::string(padSizeSuccess - (queryNameLength + padSizeQueryNumber), '.');
+    const auto queryCounterAsString = std::to_string(progressTracker.getQueryCounter() + 1);
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestRunner.cpp:328]: The thing was that with the multiple worker that we spawn now global progress tracking became necessary

File: nes-systests/systest/src/SystestParser.cpp [hunk 2]
```diff
@@ -731,6 +761,172 @@ std::string SystestParser::expectQuery()
     return queryString;
 }
 
+std::vector<ConfigurationOverride> SystestParser::expectConfiguration()
+{
+    INVARIANT(currentLine < lines.size(), "current line to parse should exist");
+    const auto& line = lines[currentLine];
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestParser.cpp:767]: The invariant above checks this already, and gives some more info in case of failure :)

File: nes-systests/systest/src/SystestStarter.cpp [hunk 3]
```diff
@@ -14,16 +14,352 @@
...
+        auto testFileDefinition = program.get<std::string>("--testLocation");
+        std::string testFilePath;
+        /// Check for test numbers (e.g., "testfile.test:5")
+        const size_t delimiterPos = testFileDefinition.find(':');
+        if (delimiterPos != std::string::npos)
+        {
+            testFilePath = testFileDefinition.substr(0, delimiterPos);
+            const std::string testNumberStr = testFileDefinition.substr(delimiterPos + 1);
+
+            std::stringstream ss(testNumberStr);
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestStarter.cpp:212]: addressed

File: nes-systests/systest/src/SystestStarter.cpp [hunk 4]
```diff
@@ -14,16 +14,352 @@
...
+            std::cout << "Cannot run systest in benchmarking mode with concurrency enabled!\n";
+            exit(-1); ///NOLINT(concurrency-mt-unsafe)
+        }
+        std::cout << "Running systests in benchmarking mode. Only one query is run at a time!\n";
+        config.numberConcurrentQueries = 1;
+    }
+
+    if (program.is_used("-d"))
+    {
+        Logger::setupLogging("systest.log", LogLevel::LOG_DEBUG);
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestStarter.cpp:188]: removed

File: nes-systests/systest/src/SystestStarter.cpp [hunk 5]
```diff
@@ -14,16 +14,352 @@
...
+    program.add_argument("--log-path").help("set the logging path");
+
+    /// debug mode
+    program.add_argument("-d", "--debug").flag().help("dump the query plan and enable debug logging");
+
+    /// input data
+    program.add_argument("--data").help("path to the directory where input CSV files are stored");
+
+    /// configs
+    program.add_argument("-w", "--workerConfig").help("load worker config file (.yaml)");
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestStarter.cpp:120]: I think so, this change just copies the options from one file to another

File: nes-systests/systest/src/SystestParser.cpp [hunk 6]
```diff
@@ -731,6 +761,172 @@ std::string SystestParser::expectQuery()
...
+        /// Check if there are commas in the value, which indicates multiple values without brackets
+        if (valueList.find(',') != std::string::npos)
+        {
+            throw SLTUnexpectedToken(
+                "Invalid configuration format for key '{}': '{}'. Multiple values must be enclosed in square brackets", key, valueList);
+        }
+        values = {valueList};
+    }
+
+    INVARIANT(!values.empty(), "when expecting a configuration keyword the configuration should not be empty");
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestParser.cpp:829]: good catch moved to exception here

File: nes-systests/systest/src/SystestParser.cpp [hunk 7]
```diff
@@ -731,6 +761,172 @@ std::string SystestParser::expectQuery()
...
+        value = NES::Util::trimWhiteSpaces(value);
+        if (value.empty())
+        {
+            throw SLTUnexpectedToken("Empty global configuration value found for key '{}'", key);
+        }
+        ConfigurationOverride override;
+        override[key] = value;
+        result.emplace_back(std::move(override));
+    }
+    return result;
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestParser.cpp:927]: addressed

File: nes-systests/systest/src/SystestBinder.cpp [hunk 8]
```diff
@@ -299,6 +312,59 @@ struct SystestBinder::Impl
     {
     }
 
+    std::vector<ConfigurationOverride>
+    mergeConfigurations(const std::vector<ConfigurationOverride>& overrides, const std::vector<ConfigurationOverride>& otherOverrides)
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/src/SystestBinder.cpp:316]: added small comment with example

File: nes-systests/systest/tests/testdata/global_config_override.dummy [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+# This query will use the global configuration
+SELECT id, value FROM stream INTO ID_VALUE_SINK
+----
+1 2
+
+# Regular configuration that will be merged with global configuration
+Configuration worker.defaultQueryExecution.pageSize3: 128
+
+# This query will also use the global configuration
+SELECT id, value FROM stream INTO ID_VALUE_SINK
```
[the remark this answers is not in the document]
Comment (adrianmichalke) [reply nes-systests/systest/tests/testdata/global_config_override.dummy:24]: The ones above (lines 6-8). Globals are for the whole file, 'locals' only the next query

**Review by ls-1801 — APPROVED** [review]
```
GlobalConfiguration worker.buffer_size_in_bytes: [4096, 8192, 16384]
GlobalConfiguration worker.default_query_execution.execution_mode: [INTERPRETER, COMPILER]

Source ...
SINK ...

Configuration worker.default_query_execution.execution_mode: [INTERPRETER]
SELECT * FROM multiBuffer INTO sinkMultiBuffer
----
```
Executes the test with `COMPILER`, I am not sure if that is the intended behavior. I would assume the perQueryConfigOverride should override the global query config. I would expect this test to be only executed 3 times.

```
  1/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=COMPILER, worker.buffer_size_in_bytes=16384]..........PASSED 
  2/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=INTERPRETER, worker.buffer_size_in_bytes=16384].......PASSED 
  3/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=COMPILER, worker.buffer_size_in_bytes=8192]...........PASSED 
  4/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=INTERPRETER, worker.buffer_size_in_bytes=8192]........PASSED 
  5/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=COMPILER, worker.buffer_size_in_bytes=4096]...........PASSED 
  6/6 BasicTypeSupport:01 [worker.default_query_execution.execution_mode=INTERPRETER, worker.buffer_size_in_bytes=4096]........PASSED 
  ```

File: nes-systests/systest/include/SystestProgressTracker.hpp [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+{
+public:
+    SystestProgressTracker();
+    explicit SystestProgressTracker(size_t totalQueries);
+
+    void incrementQueryCounter();
+    [[nodiscard]] size_t getQueryCounter() const;
+    void setTotalQueries(size_t total);
+    [[nodiscard]] size_t getTotalQueries() const;
+    [[nodiscard]] double getProgress() const;
```
Comment (yschroeder97) [inline nes-systests/systest/include/SystestProgressTracker.hpp]: Progress in percent?

File: nes-systests/systest/src/SystestExecutor.cpp [hunk 11]
```diff
@@ -62,7 +62,15 @@ using namespace std::literals;
...
-SystestConfiguration readConfiguration(int argc, const char** argv)
+SystestExecutor::SystestExecutor(SystestConfiguration config) : config(std::move(config))
+{
+}
+
+SystestExecutor::SystestExecutor(int argc, const char** argv) : SystestExecutor(parseConfiguration(argc, argv))
+{
+}
+
+SystestConfiguration SystestExecutor::parseConfiguration(int argc, const char** argv)
```
Comment (yschroeder97) [inline nes-systests/systest/src/SystestExecutor.cpp]: IMHO, parsing configuration in a type called `SystestExecutor` cries for decomposition. An executor should execute, not parse.

File: nes-systests/systest/include/SystestProgressTracker.hpp (hunk 10 above)
Comment (adrianmichalke) [reply nes-systests/systest/include/SystestProgressTracker.hpp]: addressed

**Review by yschroeder97 — APPROVED** [review]
Thanks for addressing the issues. lgtm
